docs/design/proposals/direct-blk-device-assignment.md

```diff
@@ -0,0 +1,246 @@
+container runtime, as described in [KEP-2857](https://github.com/kubernetes/enhancements/pull/2893/files). 
+However, as the KEP is still under review, we would like to propose a short/medium term solution to unblock our use case.
+
+The proposed solution is built on top of a previous [proposal](https://github.com/egernst/kata-containers/blob/da-proposal/docs/design/direct-assign-volume.md) 
+described by Eric Ernst. The previous proposal has two gaps:
+
+1. Writing a `csiPlugin.json` file to the volume root path introduces a security risk. A malicious user can gain unauthorized 
+access to a block device by writing their own `csiPlugin.json` to the above location through an ephemeral CSI plugin.  
+
+2. The proposal doesn’t describe how to establish a mapping between a volume and a kata sandbox, which is needed for 
+implementing CSI volume resize and volume stat collection APIs.
+
+This document particularly focuses on how to address these two gaps.
+
+## Assumptions
+The proposal assumes that a block device volume will only be used by one Pod on a node at a time, which we believe 
+is the more common pattern in the Kata container. It’s rare and unsafe to have the same block device attached to more than 
+one Kata Pods. In the context of Kubernetes, the PersistentVolumeClaim (PVC) needs to have accessMode as `ReadWriteOncePod`. 
+
+## End User Interface
+If a PV is intended to be consumed by a VM-based runtime, the user can specify this by adding an annotation called 
+`skip-hostmount` to the persistent volume claim (PVC). Example:
+```
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  annotations:
+    skip-hostmount: "true"
+  name: amazing-kata-volume
+spec:
+  accessModes:
+  - ReadWriteOncePod
+  resources:
+    requests:
+    storage: 4Gi
+```
+Based on this annotation, the CSI node driver knows it will not mount a filesystem on the volume and should defer the 
+operations to the container runtime. 
+
+## Interface between components
+
+Here we describe the changes we need to make in different components in order for the volume information to be propagated 
+from CSI to the Kata container runtime. 
+
+### Custom CSI node driver changes
+In the `skip-hostmount` mode, CSI node driver defers the filesystem mount operation to the container runtime. 
+The implementation of a few node driver APIs needs to be updated.
+
+* **NodePublishVolume** -- It propagates the mount information to the container runtime for it to carry out filesystem mount operation.
+* **NodeGetVolumeStats** -- Since the container runtime mounts the filesystem, it needs to forward the request to the container runtime.
+* **NodeExpandVolume** -- The CSI cannot implement this API alone. It needs to propagate the new size to the container runtime so that the filesystem metadata can be refreshed accordingly.
+* **NodeStageVolume/NodeUnStageVolume** -- Invoke a kata-runtime handler to clean up the stale mount information.
```

Comment:
Typical block device based CSI plugins (like EBS, Azure, GCE PD, Ceph, etc) will mount the PV to a global mount path in the host during NodeStage. It should be stated here that this global mount of the PV will need to be skipped during NodeStage. If the CSI plugin knows host mount should be skipped for the PV using PVC annotation or storage class parameters passed as volume context, this becomes trivial. If the CSI plugin decides to lookup pod runtimeclass to determine whether to skip host mount, then the CSI plugin should wait until NodePublishVolume to decide whether to perform/skip the host mount.